lib/client/api.go

```diff
@@ -957,8 +959,35 @@ func (tc *TeleportClient) authMethods() []ssh.AuthMethod {
 	return m
 }
 
-// ConnectToProxy dials the proxy server and returns ProxyClient if successful
-func (tc *TeleportClient) ConnectToProxy() (*ProxyClient, error) {
+// ConnectToProxy will dial to the proxy server and return a ProxyClient when
+// successful. If the passed in context is canceled, this function will return
+// a trace.ConnectionProblem right away.
+func (tc *TeleportClient) ConnectToProxy(ctx context.Context) (*ProxyClient, error) {
+	var err error
+	var proxyClient *ProxyClient
+
+	// Use connectContext and the cancel function to signal when a response is
+	// returned from connectToProxy.
+	connectContext, cancel := context.WithCancel(context.Background())
+	go func() {
+		defer cancel()
+		proxyClient, err = tc.connectToProxy()
+	}()
+
+	select {
+	// ConnectToProxy returned a result, return that back to the caller.
+	case <-connectContext.Done():
```

Comment:
I remember I left a comment earlier today about merging parent context and this one, but disregard it, your code is actually correct and my suggestion will not work.